<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function simulate GWAS marginal regression coefficients
and approximated standard error with with genotype \(\mathbf{X}\)
and phenotype \(\mathbf{y}\) simulated under the standard linear model
\(\mathbf{y} = \mathbf{X} \mathbf{b} + \mathbf{e}\). It supports
two modes: (1) simulation from the individual level genotype \(\mathbf{X}\). (2)
simulation from the moment matrix XtX: \(\mathbf{X}^T \mathbf{X}\).
In details, we want to obtain
\(\widehat{b}_j=(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}(\mathbf{X}_j^{\top}\mathbf{y})\)
and
\(\widehat{\text{s.e.}}(\widehat{b}_j) =
\widehat{\sigma_j^2}(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\).
Variance explained by each SNP is typically assumed to be small. Therefore
\(\widehat{\sigma_j^2}\) can be approximated with
\(\widehat{\sigma_j^2}=\text{Var}[\mathbf{y}]\). We further consistently
normalize \(\mathbf{y}\) such that \(\text{Var}[\mathbf{y}]=1\).
With the individual-level data \(\mathbf{X}\), we can calculate
\(\widehat{b}_j\) and \(\widehat{\text{s.e.}}[\widehat{b}_j]\) with
\((\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\) and
\((\mathbf{X}_j^{\top}\mathbf{y})\).
With the summary statistics data \(\mathbf{X}^T \mathbf{X}\), we have
$$\widehat{b}_j=
(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}(\mathbf{X}_j^{\top}\mathbf{y})=
(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}[\mathbf{X}_j^{\top}
(\mathbf{X}\mathbf{b}+\mathbf{e})]
$$
Therefore, we can simulate \(\widehat{\mathbf{b}}\) by first sample from
a multivariate normal with mean \(\mathbf{X}^\top \mathbf{X} \mathbf{b}\)
and variance \(\mathbf{X}^\top \mathbf{X} \sigma_e^2\), and then multiplied
by \((\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\) to jth entry.

"><title>Simulate GWAS from either individual level genotype or LD matrix — simulate_gwas • h2geneR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate GWAS from either individual level genotype or LD matrix — simulate_gwas"><meta property="og:description" content="This function simulate GWAS marginal regression coefficients
and approximated standard error with with genotype \(\mathbf{X}\)
and phenotype \(\mathbf{y}\) simulated under the standard linear model
\(\mathbf{y} = \mathbf{X} \mathbf{b} + \mathbf{e}\). It supports
two modes: (1) simulation from the individual level genotype \(\mathbf{X}\). (2)
simulation from the moment matrix XtX: \(\mathbf{X}^T \mathbf{X}\).
In details, we want to obtain
\(\widehat{b}_j=(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}(\mathbf{X}_j^{\top}\mathbf{y})\)
and
\(\widehat{\text{s.e.}}(\widehat{b}_j) =
\widehat{\sigma_j^2}(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\).
Variance explained by each SNP is typically assumed to be small. Therefore
\(\widehat{\sigma_j^2}\) can be approximated with
\(\widehat{\sigma_j^2}=\text{Var}[\mathbf{y}]\). We further consistently
normalize \(\mathbf{y}\) such that \(\text{Var}[\mathbf{y}]=1\).
With the individual-level data \(\mathbf{X}\), we can calculate
\(\widehat{b}_j\) and \(\widehat{\text{s.e.}}[\widehat{b}_j]\) with
\((\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\) and
\((\mathbf{X}_j^{\top}\mathbf{y})\).
With the summary statistics data \(\mathbf{X}^T \mathbf{X}\), we have
$$\widehat{b}_j=
(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}(\mathbf{X}_j^{\top}\mathbf{y})=
(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}[\mathbf{X}_j^{\top}
(\mathbf{X}\mathbf{b}+\mathbf{e})]
$$
Therefore, we can simulate \(\widehat{\mathbf{b}}\) by first sample from
a multivariate normal with mean \(\mathbf{X}^\top \mathbf{X} \mathbf{b}\)
and variance \(\mathbf{X}^\top \mathbf{X} \sigma_e^2\), and then multiplied
by \((\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\) to jth entry.

"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">h2geneR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ld.html">Example with large LD</a>
    <a class="dropdown-item" href="../articles/toy.html">Toy example</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bogdanlab/h2gene/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulate GWAS from either individual level genotype or LD matrix</h1>
      <small class="dont-index">Source: <a href="https://github.com/bogdanlab/h2gene/blob/HEAD/R/simulate.R" class="external-link"><code>R/simulate.R</code></a></small>
      <div class="d-none name"><code>simulate_gwas.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function simulate GWAS marginal regression coefficients
and <em>approximated</em> standard error with with genotype \(\mathbf{X}\)
and phenotype \(\mathbf{y}\) simulated under the standard linear model
\(\mathbf{y} = \mathbf{X} \mathbf{b} + \mathbf{e}\). It supports
two modes: (1) simulation from the individual level genotype \(\mathbf{X}\). (2)
simulation from the moment matrix <code>XtX</code>: \(\mathbf{X}^T \mathbf{X}\).</p>
<p>In details, we want to obtain
\(\widehat{b}_j=(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}(\mathbf{X}_j^{\top}\mathbf{y})\)
and
\(\widehat{\text{s.e.}}(\widehat{b}_j) =
\widehat{\sigma_j^2}(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\).</p>
<p>Variance explained by each SNP is typically assumed to be small. Therefore
\(\widehat{\sigma_j^2}\) can be approximated with
\(\widehat{\sigma_j^2}=\text{Var}[\mathbf{y}]\). We further consistently
normalize \(\mathbf{y}\) such that \(\text{Var}[\mathbf{y}]=1\).</p><ul><li><p>With the individual-level data \(\mathbf{X}\), we can calculate
\(\widehat{b}_j\) and \(\widehat{\text{s.e.}}[\widehat{b}_j]\) with
\((\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\) and
\((\mathbf{X}_j^{\top}\mathbf{y})\).</p></li>
<li><p>With the summary statistics data \(\mathbf{X}^T \mathbf{X}\), we have
$$\widehat{b}_j=
(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}(\mathbf{X}_j^{\top}\mathbf{y})=
(\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}[\mathbf{X}_j^{\top}
(\mathbf{X}\mathbf{b}+\mathbf{e})]
$$
Therefore, we can simulate \(\widehat{\mathbf{b}}\) by first sample from
a multivariate normal with mean \(\mathbf{X}^\top \mathbf{X} \mathbf{b}\)
and variance \(\mathbf{X}^\top \mathbf{X} \sigma_e^2\), and then multiplied
by \((\mathbf{X}_j^{\top}\mathbf{X}_j)^{-1}\) to jth entry.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">simulate_gwas</span><span class="op">(</span><span class="va">hsq</span>, <span class="va">beta</span>, XtX <span class="op">=</span> <span class="cn">NULL</span>, n_indiv <span class="op">=</span> <span class="cn">NULL</span>, X <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>hsq</dt>
<dd><p>Heritability explained by the genotype
\(\frac{\text{Var}[X\beta]}{\text{Var}[y]}\)</p></dd>
<dt>beta</dt>
<dd><p><code>n_snp</code> by <code>n_sim</code> simulated effect sizes</p></dd>
<dt>n_indiv</dt>
<dd><p>Number of individuals <code>n_indiv</code> for the LD matrix</p></dd>
<dt>X</dt>
<dd><p>A <code>n_indiv</code> by <code>n_snp</code> genotype matrix</p></dd>
<dt>ld</dt>
<dd><p>A <code>n_snp</code> by <code>n_snp</code> linkage disequilibrium matrix</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with the following elements:</p>
<dl><dt>beta_hat</dt>
<dd><p>Simulated marginal effects
  <code>beta_hat[, i]</code> corresponds to ith simulation</p></dd>
<dt>e</dt>
<dd><p>Simulated environmental noise</p></dd>
</dl></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Burch Kathryn, Hou Kangcheng, Ding Yi, Pasaniuc Bogdan.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer></div>

  

  

  </body></html>

